pkg/store/proxy.go

```diff
@@ -0,0 +1,242 @@
+package store
+
+import (
+	"context"
+	"sync"
+
+	"io"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/query"
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/improbable-eng/thanos/pkg/strutil"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb/labels"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// ProxyStore implements the store API that proxies request to all given underlying stores.
+type ProxyStore struct {
+	logger         log.Logger
+	stores         func() []*query.StoreInfo
+	selectorLabels labels.Labels
+}
+
+// NewProxyStore returns a new ProxyStore that uses the given clients that implements storeAPI to fan-in all series to the client.
+// Note that there is no deduplication support. Deduplication should be done on the highest level (just before PromQL)
+func NewProxyStore(
+	logger log.Logger,
+	stores func() []*query.StoreInfo,
+	selectorLabels labels.Labels,
+) *ProxyStore {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	s := &ProxyStore{
+		logger:         logger,
+		stores:         stores,
+		selectorLabels: selectorLabels,
+	}
+	return s
+}
+
+// Info returns store information about the external labels this store have.
+func (s *ProxyStore) Info(ctx context.Context, r *storepb.InfoRequest) (*storepb.InfoResponse, error) {
+	res := &storepb.InfoResponse{
+		Labels: make([]storepb.Label, 0, len(s.selectorLabels)),
+	}
+	for _, l := range s.selectorLabels {
+		res.Labels = append(res.Labels, storepb.Label{
+			Name:  l.Name,
+			Value: l.Value,
+		})
+	}
+	return res, nil
+}
+
+// Series returns all series for a requested time range and label matcher. Requested series are taken from other
+// stores and proxied to RPC client. NOTE: Resulted data are not trimmed exactly to min and max time range.
+func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	match, newMatchers, err := labelsMatches(s.selectorLabels, r.Matchers)
+	if err != nil {
+		return status.Error(codes.InvalidArgument, err.Error())
+	}
+	if !match {
+		return nil
+	}
+
+	var (
+		respCh    = make(chan *storepb.SeriesResponse, 1024)
```

Comment:
That seems like an excessive buffer size. We just want enough so the merging iterator always has some data available as it calls `Next()`. 1024 is about the maximum size of series we expect a query to consider, which means you'll probably always load all data into the buffer here, which inflates memory usage and defeats the purpose of the streaming API.

Back pressure is completely fine – if the consumer doesn't read new series fast enough, why should we load them all into memory in the mean time?

Reply:
I'd probably make this 5 or 10 at max. Especially because some series can be huge.

Reply:
I just noticed that this is actually the channel for the merged series. This channel should just be removed entirely . You can write straight from the merged set into the output. As each series set already has its own buffer, this doesn't add anything. Popping the next element from the merged set is O(microseconds) and thus nothing worth buffering up.

Reply:
So, I think all you said would be true, if there were no partial error warnings that are required to be set outside of the mergedSet series pop (: Your comments regarding buffer size are valid, though - will change to smaller value for both `respCh` and `recvCh`